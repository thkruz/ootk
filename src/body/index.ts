export { Celestial } from './Celestial';
export { Earth } from './Earth';
export { Moon } from './Moon';
export { Sun } from './Sun';
