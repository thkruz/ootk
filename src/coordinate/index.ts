export * from './Hill.js';
export * from './ClassicalElements.js';
export * from './EquinoctialElements.js';
export * from './FormatTle.js';
export * from './Geodetic.js';
export * from './ITRF.js';
export * from './J2000.js';
export * from './RelativeState.js';
export * from './RIC.js';
export * from './StateVector.js';
export * from './TEME.js';
export * from './Tle.js';
