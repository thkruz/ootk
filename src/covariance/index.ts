export * from './CovarianceSample.js';
export * from './StateCovariance.js';
