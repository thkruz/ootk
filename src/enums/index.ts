export { CatalogSource } from './CatalogSource.js';
export { CommLink } from './CommLink.js';
export { OrbitRegime } from './OrbitRegime.js';
export { Sgp4OpsMode } from './Sgp4OpsMode.js';
export { AngularDiameterMethod } from './AngularDiameterMethod.js';
export { AngularDistanceMethod } from './AngularDistanceMethod.js';
export { PassType } from './PassType.js';
