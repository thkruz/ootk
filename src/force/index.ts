export { EarthGravity } from './EarthGravity.js';
export { Force } from './Force.js';
export { ForceModel } from './ForceModel.js';
export { Gravity } from './Gravity.js';
export { SolarRadiationPressure } from './SolarRadiationPressure.js';
export { ThirdBodyGravity } from './ThirdBodyGravity.js';
export { Thrust } from './Thrust.js';

