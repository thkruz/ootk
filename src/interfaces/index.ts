export type { GroundPositionParams } from './GroundPositionParams.js';
export type { OptionsParams } from './OptionsParams.js';
export type { TimeVariables } from './TimeVariables.js';
export type { StarObjectParams } from './StarObjectParams.js';
export type { ClassicalElementsParams } from './ClassicalElementsParams.js';
export type { EquinoctialElementsParams } from './EquinoctialElementsParams.js';
export type { ScenarioParams } from '../objects/Scenario.js';
export type { BaseObjectParams } from '../objects/base/BaseObject.js';
export type { GroundObjectParams } from '../objects/base/GroundObject.js';
export type { FacilityParams } from '../objects/base/Facility.js';
export type { SensorParams } from '../objects/attachable/SensorAdv.js';
export type { RectangularSensorParams } from '../objects/attachable/RectangularSensor.js';
