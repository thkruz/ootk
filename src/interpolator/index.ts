export { ChebyshevCoefficients } from './ChebyshevCoefficients.js';
export { ChebyshevInterpolator } from './ChebyshevInterpolator.js';
export { CubicSpline } from './CubicSpline.js';
export { CubicSplineInterpolator } from './CubicSplineInterpolator.js';
export { Interpolator } from './Interpolator.js';
export { LagrangeInterpolator } from './LagrangeInterpolator.js';
export { StateInterpolator } from './StateInterpolator.js';
export { VerletBlendInterpolator } from './VerletBlendInterpolator.js';

