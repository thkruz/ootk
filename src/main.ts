/**
 * @author @thkruz Theodore Kruczek
 * @description Orbital Object ToolKit (ootk) is a collection of tools for working
 * with satellites and other orbital objects.
 * @license AGPL-3.0-or-later
 * @copyright (c) 2020-2024 Theodore Kruczek
 *
 * Many of the classes are based off of the work of @david-rc-dayton and his
 * Pious Squid library (https://github.com/david-rc-dayton/pious_squid).
 *
 * Orbital Object ToolKit is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * Orbital Object ToolKit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with
 * Orbital Object ToolKit. If not, see <http://www.gnu.org/licenses/>.
 */

export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './time/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
export * from './operations/operations.js';
export { BaseObject, Sensor } from './objects/index.js';
export { Earth } from './body/index.js';
export * from './coordinate/index.js';
export * from './observation/index.js';
export * from './data/DataHandler.js';
export * from './sgp4/index.js';
export * from './objects/index.js';
export * from './interpolator/index.js';
export * from './force/index.js';
export * from './maneuver/index.js';
export * from './optimize/index.js';
export * from './orbit_determination/index.js';
export * from './body/index.js';
export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './coordinate/index.js';
export * from './operations/index.js';
export * from './force/index.js';
export * from './propagator/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
