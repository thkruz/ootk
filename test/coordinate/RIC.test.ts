import { EpochUTC, J2000, Kilometers, KilometersPerSecond, Matrix, RIC, Vector3D } from './../../src/main';

describe('RIC', () => {
  let stateVector: J2000;
  let origin: J2000;

  beforeEach(() => {
    stateVector = new J2000(
      EpochUTC.fromDateTime(new Date(1705109326817)),
      new Vector3D(1538.223335842895 as Kilometers, 5102.261204021967 as Kilometers, 4432.634965003577 as Kilometers),
      new Vector3D(
        -7.341518909379302 as KilometersPerSecond,
        0.6516718453998644 as KilometersPerSecond,
        1.7933882499861993 as KilometersPerSecond,
      ),
    );

    origin = new J2000(
      EpochUTC.fromDateTime(new Date(1705109326817)),
      new Vector3D(1540.223335842895 as Kilometers, 5102.251204021967 as Kilometers, 4432.644965003577 as Kilometers),
      new Vector3D(
        -7.331518909379302 as KilometersPerSecond,
        0.6516718453998644 as KilometersPerSecond,
        1.7933882499861993 as KilometersPerSecond,
      ),
    );
  });

  // Create a new RIC coordinate from J2000 state vectors and origin
  it('should create a new RIC coordinate from J2000 state vectors and origin', () => {
    const transform = new Matrix([
      [1, 0, 0],
      [0, 1, 0],
      [0, 0, 1],
    ]);

    const ric = RIC.fromJ2000Matrix(stateVector, origin, transform);

    expect(ric).toMatchSnapshot();
  });

  // Create a RIC coordinate system from a J2000 state and origin
  it('should create a RIC coordinate system from a J2000 state and origin', () => {
    const ric = RIC.fromJ2000(stateVector, origin);

    expect(ric).toMatchSnapshot();
  });

  /*
   * Transform the current RIC coordinate to the J2000 coordinate system using
   * the provided origin
   */
  it('should transform the current RIC coordinate to the J2000 coordinate system using the provided origin', () => {
    const transform = new Matrix([
      [1, 0, 0],
      [0, 1, 0],
      [0, 0, 1],
    ]);
    const ric = new RIC(stateVector.position, stateVector.velocity);
    const j2000 = ric.toJ2000Matrix(origin, transform);

    expect(j2000).toMatchSnapshot();
  });

  /*
   * Transform the current RIC coordinate to the J2000 coordinate system using
   * the provided origin and transform matrix
   */
  it('should transform the current RIC coordinate to J2000 using the provided origin and transform matrix', () => {
    const ric = new RIC(stateVector.position, stateVector.velocity);
    const j2000 = ric.toJ2000(origin);

    expect(j2000).toMatchSnapshot();
  });

  // Get the name of the RIC coordinate system
  it('should get the name of the RIC coordinate system', () => {
    const ric = new RIC(stateVector.position, stateVector.velocity);
    const name = ric.name;

    expect(name).toMatchSnapshot();
  });

  // toString
  it('should return a string representation of the RIC coordinate', () => {
    const ric = new RIC(stateVector.position, stateVector.velocity);
    const string = ric.toString();

    expect(string).toMatchSnapshot();
  });

  // range
  it('should return the range of the RIC coordinate', () => {
    const ric = new RIC(stateVector.position, stateVector.velocity);
    const range = ric.range;

    expect(range).toMatchSnapshot();
  });

  // rangeRate
  it('should return the range rate of the RIC coordinate', () => {
    const ric = new RIC(stateVector.position, stateVector.velocity);
    const rangeRate = ric.rangeRate;

    expect(rangeRate).toMatchSnapshot();
  });
});
