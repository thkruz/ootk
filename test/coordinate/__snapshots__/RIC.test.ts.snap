// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RIC should create a RIC coordinate system from a J2000 state and origin 1`] = `
RIC {
  "position": Vector3D {
    "x": -0.4434101992280181,
    "y": 1.9338646031157782,
    "z": -0.25249770699242563,
  },
  "velocity": Vector3D {
    "x": -0.002221880767217895,
    "y": 0.009676856714746954,
    "z": -0.001192346417169753,
  },
}
`;

exports[`RIC should create a new RIC coordinate from J2000 state vectors and origin 1`] = `
RIC {
  "position": Vector3D {
    "x": -2,
    "y": 0.010000000000218279,
    "z": -0.010000000000218279,
  },
  "velocity": Vector3D {
    "x": -0.010000000000000675,
    "y": 0,
    "z": 0,
  },
}
`;

exports[`RIC should get the name of the RIC coordinate system 1`] = `"RIC"`;

exports[`RIC should return a string representation of the RIC coordinate 1`] = `
"[RIC]
  Position: [1538.223336, 5102.261204, 4432.634965] km
  Velocity: [-7.341518909, 0.651671845, 1.793388250] km/s"
`;

exports[`RIC should return the range of the RIC coordinate 1`] = `6931.627020979389`;

exports[`RIC should return the range rate of the RIC coordinate 1`] = `-0.002663194316531782`;

exports[`RIC should transform the current RIC coordinate to J2000 using the provided origin and transform matrix 1`] = `
J2000 {
  "epoch": EpochUTC {
    "posix": 1705109326.817,
  },
  "position": Vector3D {
    "x": -2526.8631969362687,
    "y": 3698.432701341307,
    "z": 9867.307227222767,
  },
  "velocity": Vector3D {
    "x": -9.379496377921184,
    "y": -5.899851460084735,
    "z": -1.4349469513344795,
  },
}
`;

exports[`RIC should transform the current RIC coordinate to the J2000 coordinate system using the provided origin 1`] = `
J2000 {
  "epoch": EpochUTC {
    "posix": 1705109326.817,
  },
  "position": Vector3D {
    "x": 3078.44667168579,
    "y": 10204.512408043935,
    "z": 8865.279930007155,
  },
  "velocity": Vector3D {
    "x": -14.673037818758605,
    "y": 1.303343690799729,
    "z": 3.5867764999723986,
  },
}
`;
